[BUG] regression - npm_config_prefix no longer works with global npx

### Is there an existing issue for this?

- [X] I have searched the existing issues

### This issue exists in the latest npm version

- [X] I am using the latest npm

### Current Behavior

Using the `npm_config_prefix` results in a `no such file or directory` error when attempting to run a command via `npx` globally on `npm@8.16.0`. The same command used to work on `npm@8.15.1`.

### Expected Behavior

Global npx command works in conjunction with `npm_config_prefix` outside of a project/workspace.

### Steps To Reproduce

```
$ mkdir repro
$ cd repro
$ npm i -g npm@8.15.1
$ npm_config_prefix=/tmp/.foo npx serve@latest --help
# observe `serve` CLI is executed successfully and the help message is printed
```

Now, upgrade `npm` and repeat:

```
$ npm i -g npm@8.16.0
$ npm_config_prefix=/tmp/.foo npx serve@latest --help
npm ERR! code ENOENT
npm ERR! syscall lstat
npm ERR! path /private/tmp/.foo/lib
npm ERR! errno -2
npm ERR! enoent ENOENT: no such file or directory, lstat '/private/tmp/.foo/lib'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent 

npm ERR! A complete log of this run can be found in:
npm ERR!     /…/.npm/_logs/2022-08-05T21_22_57_389Z-debug-0.log
```

💥 Observe `npx` errors and complains `This is related to npm not being able to find a file.`.

### Environment

- npm:
  - GOOD: 8.15.1
  - BAD: 8.16.0
- Node.js: v16.14.0
- OS Name: macOS, Ubuntu 22
- System Model Name: M1 Macbook Pro, GitHub Action Runner
- npm config:
```ini
$ npm config ls
; node bin location = /Users/…/.nvm/versions/node/v16.14.0/bin/node
; node version = v16.14.0
; npm local prefix = /Users/…/Desktop/repro
; npm version = 8.16.0
; cwd = /Users/…/Desktop/repro
; HOME = /Users/…
; Run `npm config ls -l` to show all defaults.
```

Additionally, `npm_config_prefix` is set when running `npx` command.

@ljharb It looks like that gives the same error as it's [equivalent to the env var](https://docs.npmjs.com/cli/v8/using-npm/config#environment-variables):

```
$ npm config --global set prefix /tmp/bar
$ npx serve
npm ERR! code ENOENT
npm ERR! syscall lstat
npm ERR! path /private/tmp/bar
npm ERR! errno -2
npm ERR! enoent ENOENT: no such file or directory, lstat '/private/tmp/bar'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent 

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/…/.npm/_logs/2022-08-09T04_50_53_983Z-debug-0.log
```

Even doing:

```
$ npx --prefix=/tmp/bar serve
```

produces the same error (with fresh install/config). Given the above, it looks like with 8.16.0 you can no longer use a non-default global install directory.

--- a/lib/commands/exec.js
+++ b/lib/commands/exec.js
@@ -34,6 +34,7 @@ class Exec extends BaseCommand {
 
     const args = [..._args]
     const call = this.npm.config.get('call')
+    let globalPath
     const {
       flatOptions,
       localBin,
@@ -44,6 +45,12 @@ class Exec extends BaseCommand {
     const scriptShell = this.npm.config.get('script-shell') || undefined
     const packages = this.npm.config.get('package')
     const yes = this.npm.config.get('yes')
+    // --prefix sets both of these to the same thing, meaning the global prefix
+    // is invalid (i.e. no lib/node_modules).  This is not a trivial thing to
+    // untangle and fix so we work around it here.
+    if (this.npm.localPrefix !== this.npm.globalPrefix) {
+      globalPath = path.resolve(globalDir, '..')
+    }
 
     if (call && _args.length) {
       throw this.usageError()
@@ -59,7 +66,7 @@ class Exec extends BaseCommand {
       localBin,
       locationMsg,
       globalBin,
-      globalPath: path.resolve(globalDir, '..'),
+      globalPath,
       output,
       packages,
       path: localPrefix,

--- a/test/lib/commands/exec.js
+++ b/test/lib/commands/exec.js
@@ -28,6 +28,7 @@ t.test('registry package', async t => {
 
   const { npm } = await loadMockNpm(t, {
     config: {
+      audit: false,
       yes: true,
     },
     prefixDir: {
@@ -43,7 +44,46 @@ t.test('registry package', async t => {
   })
 
   await registry.package({
-    times: 2,
+    manifest,
+    tarballs: {
+      '1.0.0': path.join(npm.prefix, 'npm-exec-test'),
+    } })
+
+  await npm.exec('exec', ['@npmcli/npx-test'])
+  const exists = await fs.stat(path.join(npm.prefix, 'npm-exec-test-success'))
+  t.ok(exists.isFile(), 'bin ran, creating file')
+})
+
+t.test('--prefix', async t => {
+  const registry = new MockRegistry({
+    tap: t,
+    registry: 'https://registry.npmjs.org/',
+  })
+
+  const manifest = registry.manifest({ name: '@npmcli/npx-test' })
+  manifest.versions['1.0.0'].bin = { 'npx-test': 'index.js' }
+
+  const { npm } = await loadMockNpm(t, {
+    config: {
+      audit: false,
+      yes: true,
+    },
+    prefixDir: {
+      'npm-exec-test': {
+        'package.json': JSON.stringify(manifest),
+        'index.js': `#!/usr/bin/env node
+  require('fs').writeFileSync('npm-exec-test-success', '')`,
+      },
+    },
+    globals: ({ prefix }) => ({
+      'process.cwd': () => prefix,
+    }),
+  })
+
+  // This is what `--prefix` does
+  npm.globalPrefix = npm.localPrefix
+
+  await registry.package({
     manifest,
     tarballs: {
       '1.0.0': path.join(npm.prefix, 'npm-exec-test'),
@@ -65,6 +105,7 @@ t.test('workspaces', async t => {
 
   const { npm } = await loadMockNpm(t, {
     config: {
+      audit: false,
       yes: true,
       workspace: ['workspace-a'],
     },

--- a/workspaces/libnpmexec/lib/index.js
+++ b/workspaces/libnpmexec/lib/index.js
@@ -82,7 +82,7 @@ const exec = async (opts) => {
     localBin = resolve('./node_modules/.bin'),
     locationMsg = undefined,
     globalBin = '',
-    globalPath = '',
+    globalPath,
     output,
     // dereference values because we manipulate it later
     packages: [...packages] = [],
@@ -120,7 +120,7 @@ const exec = async (opts) => {
     if (localBinPath) {
       binPaths.push(localBinPath)
       return await run()
-    } else if (await fileExists(`${globalBin}/${args[0]}`)) {
+    } else if (globalPath && await fileExists(`${globalBin}/${args[0]}`)) {
       binPaths.push(globalBin)
       return await run()
     }
@@ -155,7 +155,7 @@ const exec = async (opts) => {
 
     args[0] = getBinFromManifest(commandManifest)
 
-    if (needInstall.length > 0) {
+    if (needInstall.length > 0 && globalPath) {
       // See if the package is installed globally, and run the translated bin
       const globalArb = new Arborist({ ...flatOptions, path: globalPath, global: true })
       const globalTree = await globalArb.loadActual()

--- a/workspaces/libnpmexec/test/index.js
+++ b/workspaces/libnpmexec/test/index.js
@@ -517,6 +517,7 @@ t.test('global space pkg', async t => {
     },
   })
   const globalBin = resolve(path, 'global/node_modules/.bin')
+  const globalPath = resolve(path, 'global')
   const runPath = path
 
   const executable = resolve(path, 'global/node_modules/a')
@@ -531,6 +532,7 @@ t.test('global space pkg', async t => {
     ...baseOpts,
     args: ['a', 'resfile'],
     globalBin,
+    globalPath,
     path,
     runPath,
   })
@@ -588,12 +590,13 @@ t.test('run from registry - no local packages', async t => {
   const testdir = t.testdir({
     cache: {},
     npxCache: {},
+    global: {
+      lib: {},
+      bin: {},
+    },
     work: {},
   })
   const path = resolve(testdir, 'work')
-  const runPath = path
-  const cache = resolve(testdir, 'cache')
-  const npxCache = resolve(testdir, 'npxCache')
 
   t.throws(
     () => fs.statSync(resolve(path, 'index.js')),
@@ -604,10 +607,11 @@ t.test('run from registry - no local packages', async t => {
   await libexec({
     ...baseOpts,
     args: ['@ruyadorno/create-index'],
-    cache,
-    npxCache,
+    cache: resolve(testdir, 'cache'),
+    globalPath: resolve(testdir, 'global'),
+    npxCache: resolve(testdir, 'npxCache'),
     path,
-    runPath,
+    runPath: path,
   })
 
   t.ok(fs.statSync(resolve(path, 'index.js')).isFile(), 'ran create pkg')

